File: projects/Ruby/benchmarks/benchmark_serialization_final.rb
require 'benchmark'

require_relative '../proto/proto'

class BenchmarkSerializationFinal
  def initialize
    # Create a new account with some orders
    @account = Proto::Account.new(1, 'Test', Proto::State.good, Proto::Balance.new('USD', 1000.0), Proto::Balance.new('EUR', 100.0))
    @account.orders.push(Proto::Order.new(1, 'EURUSD', Proto::OrderSide.buy, Proto::OrderType.market, 1.23456, 1000.0))
    @account.orders.push(Proto::Order.new(2, 'EURUSD', Proto::OrderSide.sell, Proto::OrderType.limit, 1.0, 100.0))
    @account.orders.push(Proto::Order.new(3, 'EURUSD', Proto::OrderSide.buy, Proto::OrderType.stop, 1.5, 10.0))

    # Serialize the account to the FBE stream
    @writer = Proto::AccountFinalModel.new(FBE::WriteBuffer.new)
    @writer.serialize(@account)
    raise "Verify error!" unless @writer.verify

    # Deserialize the account from the FBE stream
    @reader = Proto::AccountFinalModel.new(FBE::ReadBuffer.new)
    @reader.attach_buffer(@writer.buffer)
    raise "Verify error!" unless @reader.verify
    @reader.deserialize(@account)
  end

  def serialize
    # Reset FBE stream
    @writer.reset

    # Serialize the account to the FBE stream
    @writer.serialize(@account)
  end

  def deserialize
    # Deserialize the account from the FBE stream
    @reader.deserialize(@account)
  end

  def verify
    # Verify the account
    @writer.verify
  end
end

benchmark = BenchmarkSerializationFinal.new
iterations = 100000

def report(name, duration, iterations)
  puts
  puts "Phase: #{name}"
  puts "Average time: %.3f mcs / iteration" % (duration / iterations * 1000000)
  puts "Total time: %.3f s" % duration
  puts "Total iterations: %d" % iterations
  puts "Iterations throughput: %.3f / second" % (iterations / duration)
  puts
end

puts "Benchmarking verify() method..."
duration = Benchmark.realtime do
  iterations.times { benchmark.verify }
end
report("Validate (Final)", duration, iterations)

puts "Benchmarking serialize() method..."
duration = Benchmark.realtime do
  iterations.times { benchmark.serialize }
end
report("Serialize (Final)", duration, iterations)

puts "Benchmarking deserialize() method..."
duration = Benchmark.realtime do
  iterations.times { benchmark.deserialize }
end
report("Deserialize (Final)", duration, iterations)

File: projects/Python/benchmarks/benchmark_serialization_final.py
import fbe
import proto
import timeit
from proto import proto


class BenchmarkSerializationFinal(object):

    def __init__(self):
        # Create a new account with some orders
        self.account = proto.Account(1, "Test", proto.State.good, proto.Balance("USD", 1000.0), proto.Balance("EUR", 100.0))
        self.account.orders.append(proto.Order(1, "EURUSD", proto.OrderSide.buy, proto.OrderType.market, 1.23456, 1000.0))
        self.account.orders.append(proto.Order(2, "EURUSD", proto.OrderSide.sell, proto.OrderType.limit, 1.0, 100.0))
        self.account.orders.append(proto.Order(3, "EURUSD", proto.OrderSide.buy, proto.OrderType.stop, 1.5, 10.0))

        # Serialize the account to the FBE stream
        self.writer = proto.AccountFinalModel(fbe.WriteBuffer())
        self.writer.serialize(self.account)
        assert self.writer.verify()

        # Deserialize the account from the FBE stream
        self.reader = proto.AccountFinalModel(fbe.ReadBuffer())
        self.reader.attach_buffer(self.writer.buffer)
        assert self.reader.verify()
        self.reader.deserialize(self.account)

    def serialize(self):
        # Reset FBE stream
        self.writer.reset()

        # Serialize the account to the FBE stream
        self.writer.serialize(self.account)

    def deserialize(self):
        # Deserialize the account from the FBE stream
        self.reader.deserialize(self.account)

    def verify(self):
        # Verify the account
        self.writer.verify()


benchmark = BenchmarkSerializationFinal()
iterations = 100000


def report(name, duration):
    print()
    print("Phase: {}".format(name))
    print("Average time: {:.3f} mcs / iteration".format(duration / iterations * 1000000))
    print("Total time: {:.3f} s".format(duration))
    print("Total iterations: {}".format(iterations))
    print("Iterations throughput: {:.3f} / second".format(iterations / duration))
    print()


def main():
    # Benchmark verify() method
    times = timeit.repeat(setup='print("Benchmarking verify() method...")', stmt="benchmark.verify()", repeat=5, number=iterations, globals=globals())
    duration = min(times)
    report("Verify (Final)", duration)

    # Benchmark serialize() method
    times = timeit.repeat(setup='print("Benchmarking serialize() method...")', stmt="benchmark.serialize()", repeat=5, number=iterations, globals=globals())
    duration = min(times)
    report("Serialize (Final)", duration)

    # Benchmark deserialize() method
    times = timeit.repeat(setup='print("Benchmarking deserialize() method...")', stmt="benchmark.deserialize()", repeat=5, number=iterations, globals=globals())
    duration = min(times)
    report("Deserialize (Final)", duration)


if __name__ == "__main__":
    main()
